reproman/interface/docker.shim

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/perl
+#emacs: -*- mode: shell-script; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: t -*-
+#ex: set sts=4 ts=4 sw=4 noet:
+#
+#
+# This script is a shim for the docker executable. The shim intercepts
+# calls to the docker executable, runs a image id discovery process, and
+# then proceeds with the original call to the original docker executable.
+#
+require 5.002;
+
+# Print debugging if docker debug switch is set
+my $DEBUG = 0;
+foreach(@ARGV) {
+    $DEBUG = 1 if ($_ eq '--debug' or $_ eq '-D');
+}
+
+# Print debug output
+# 
+# Parameters
+# ----------
+# $output
+#     string : Text to print to the screen.
+# 
+sub debug {
+    return if (!$DEBUG);
+    foreach my $line (@_) {
+        print "[DEBUG DOCKER SHIM] ${line}";
+        print "\n" if (!($line =~ /\n$/));
+    }
+}
+
+# Run commands in shell
+# 
+# Parameters
+# ----------
+# $command
+#     string : Command to execute in the shell
+# $default_message
+#     string : Message to send if nothing is found in command stderr
+#
+# Returns
+# -------
+# string : stdout of the executed command
+# 
+sub exec_handler {
+
+	my $command = shift;
+	my $default_message = shift;
+
+	my $output = qx/$command 2>&1/;
+	if ($? != 0) {
+		if ($output) {
+			die $output
+		}
+		else {
+  		die $default_message;
+		}
+	}
+	return $output
+}
+
+# Gather ENV info
+my $TRACERS_DIR = $ENV{"REPROMAN_TRACERS_DIR"};
+if (!$TRACERS_DIR) {
+	die "ENV variable REPROMAN_TRACERS_DIR not found";
+}
+debug("TRACERS_DIR: ${TRACERS_DIR}");
+my $EXTRA_TRACE_FILE = $ENV{"REPROMAN_EXTRA_TRACE_FILE"};
+if (!$EXTRA_TRACE_FILE) {
+	die "ENV variable REPROMAN_EXTRA_TRACE_FILE not found";
+}
+debug("EXTRA_TRACE_FILE: ${EXTRA_TRACE_FILE}");
+my $WAIT_SCRIPT = "wait.sh";
+
+# Locate original Docker executable (Most likely in /usr/bin)
+my $stdout = exec_handler("which -a docker", "Unable to execute system 'which' command");
+my @paths = split(/\s+/, $stdout);
+my $original_docker_exec = '';
+for my $path (@paths) {
+  $original_docker_exec = $path if $path ne $0;
+}
+if (!$original_docker_exec) {
+  die "Original docker executable not found";
+}
+debug("Found docker executable: ${original_docker_exec}");
+
+# Create the script the will be the entrypoint of the container needed to
+# figure out the image RepoDigest value. The container will be removed when no
+# longer needed or, if the shim fails, will remove itself after the sleep
+# command completes.
+open(my $fh, '>', "${TRACERS_DIR}/${WAIT_SCRIPT}")
+  or die "Could not open file '${TRACERS_DIR}/${WAIT_SCRIPT}' $!";
+print $fh <<'EOF';
+#!/bin/sh
+sleep 30
+EOF
+close($fh);
+chmod(0755, "${TRACERS_DIR}/${WAIT_SCRIPT}");
+
+# Pre-run the docker container in detached mode to get its id
+my $cmd = "${original_docker_exec} run -d --rm ";
+$cmd .= "-v ${TRACERS_DIR}/${WAIT_SCRIPT}:/opt/${WAIT_SCRIPT} ";
+$cmd .= "--entrypoint /opt/${WAIT_SCRIPT} ";
+my ($start_index) = grep { $ARGV[$_] eq 'run' } (0 .. @ARGV-1);
```

Comment:
As mentioned above, we should die appropriately when "run" isn't given.  Also I think something like

```perl
List::Util qw(first);
first { $ARGV[$_] eq 'run' } 0..$#ARGV;
```

would be more direct way to get the index.